From afb6b2d59ae4398dd4841b9fc34aa187cf3a6906 Mon Sep 17 00:00:00 2001
From: Dave Young <dyoung@redhat.com>
Date: Thu, 12 Jun 2014 02:34:05 -0400
Subject: [x86] efi: Do not export efi runtime map in case old map

Message-id: <20140612023405.GA1225@dhcp-16-198.nay.redhat.com>
Patchwork-id: 83686
O-Subject: [RHEL7.1 PATCH 37/35] x86/efi: Do not export efi runtime map in case old map
Bugzilla: 1080109
RH-Acked-by: Chao WANG <chaowang@redhat.com>
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

There's a remain issue to support efi old_map, with old_map kernel will use
ioremap for efi memory mapping instead of new 1:1 mapping.
In this case kexec/kdump need switch to old way to load kernels.

Resolves: bz1080109

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7553657

Test: Tested kexec/kdump on HP workstation with efi=old_map.
Test kexec with extra acpi_rsdp cmdline. Kdump works automaticlly because
kexec tools will add acpi_rsdp cmdline for us.

Upstream commit:
commit a3530e8fe980f756b823d451fe9243f26db95fa5
Author: Dave Young <dyoung@redhat.com>
Date:   Fri May 30 11:20:10 2014 +0800

    x86/efi: Do not export efi runtime map in case old map

    For ioremapped efi memory aka old_map the virt addresses are not persistant
    across kexec reboot. kexec-tools will read the runtime maps from sysfs then
    pass them to 2nd kernel and assuming kexec efi boot is ok. This will cause
    kexec boot failure.

    To address this issue do not export runtime maps in case efi old_map so
    userspace can use no efi boot instead.

    Signed-off-by: Dave Young <dyoung@redhat.com>
    Acked-by: Borislav Petkov <bp@suse.de>
    Acked-by: Vivek Goyal <vgoyal@redhat.com>
    Signed-off-by: Matt Fleming <matt.fleming@intel.com>
---
This patch is in mainline 3.15 kernel, it's on top of previous 36 patches for RHEL7.1
If need I repost the whole patchset, please let me know. Thanks a lot.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/platform/efi/efi.c b/arch/x86/platform/efi/efi.c
index 3c90af9..574b2a7 100644
--- a/arch/x86/platform/efi/efi.c
+++ b/arch/x86/platform/efi/efi.c
@@ -902,6 +902,9 @@ static void __init save_runtime_map(void)
  void *tmp, *p, *q = NULL;
  int count = 0;
 
+ if (efi_enabled(EFI_OLD_MEMMAP))
+  return;
+
  for (p = memmap.map; p < memmap.map_end; p += memmap.desc_size) {
   md = p;
 
-- 
1.7.1